src/garage/experiment/local_tf_runner.py

```diff
@@ -340,73 +377,36 @@ def resume(self,
-            start_epoch=self.train_args.start_epoch)
-
-    def _train(self,
-               n_epochs,
-               n_epoch_cycles,
-               batch_size,
-               plot,
-               store_paths,
-               pause_for_plot,
-               start_epoch=0):
+        if n_epochs:
+            self.train_args.n_epochs = n_epochs
+        if batch_size:
+            self.train_args.batch_size = batch_size
+        if n_epoch_cycles:
+            self.train_args.n_epoch_cycles = n_epoch_cycles
+        if plot:
+            self.train_args.plot = plot
+        if store_paths:
+            self.train_args.store_paths = store_paths
+        if pause_for_plot:
+            self.train_args.pause_for_plot = pause_for_plot
+
+        return self._train()
+
+    def _train(self):
         """Start actual training.
 
-        Args:
-            n_epochs(int): Number of epochs.
-            n_epoch_cycles(int): Number of batches of samples in each epoch.
-                This is only useful for off-policy algorithm.
-                For on-policy algorithm this value should always be 1.
-            batch_size(int): Number of steps in batch.
-            plot(bool): Visualize policy by doing rollout after each epoch.
-            store_paths(bool): Save paths in snapshot.
-            pause_for_plot(bool): Pause for plot.
-            start_epoch: (internal) The starting epoch.
-                Use for experiment resuming.
-
         Returns:
             The average return in last epoch cycle.
 
         """
-        assert self.has_setup, ('Use Runner.setup() to setup runner before '
-                                'training.')
-
-        # Save arguments for restore
-        self.train_args = SimpleNamespace(
-            n_epochs=n_epochs,
-            n_epoch_cycles=n_epoch_cycles,
-            batch_size=batch_size,
-            plot=plot,
-            store_paths=store_paths,
-            pause_for_plot=pause_for_plot,
-            start_epoch=start_epoch)
-
-        self.start_worker()
-
-        self.start_time = time.time()
-        itr = start_epoch * n_epoch_cycles
 
         last_return = None
-        for epoch in range(start_epoch, n_epochs):
-            self.itr_start_time = time.time()
-            paths = None
-            with logger.prefix('epoch #%d | ' % epoch):
-                for cycle in range(n_epoch_cycles):
-                    paths = self.obtain_samples(itr, batch_size)
-                    last_return = self.algo.train_once(itr, paths)
-                    itr += 1
-                self.save(epoch, paths if store_paths else None)
-                self.log_diagnostics(pause_for_plot)
-                logger.dump_all(itr)
-                tabular.clear()
-
-        self.shutdown_worker()
+        for _ in self._step_epochs():
+            for cycle in range(self.train_args.n_epoch_cycles):
+                self.step_epochs_args.path = self.obtain_samples(
```

Comment:
It's true that `path` (and `itr`) are used by `step_epochs()`, but using a namespace is fairly overkill, and feels really strange.

Please just make `path` and `itr` fields of the runner instead.

@ryanjulian Do you have other opinions on how to fix this API? Maybe these should be fields on the Algorithm?

Reply:
i was just going to comment on this. the `SimpleNamespace` was intended only for convenient storage (pickling) and passing back to _train(), not as objects fields. it should never be used as a container for object fields, so I think we should refactor this back to use only fields.

as for `path` and `itr`: why is `path` saved anyway? the return from `obtain_samples` could be Very Large, and we don't necessarily want to save it.

i'm not really offended by `itr` because the runner needs that information (e.g. for logging, saving snapshots). it should be a field of the runner and it will also be a field in the experiment pkl file.

Reply:
Sometimes it's useful to store `path`. Right now, `path` is only saved if `store_paths == True` (it defaults to `False`), which addresses the concern with it potentially being Very Large. It should probably be on the same object as `itr`? 

Alternatively, we could have some mechanism for explicitly passing paths to be saved, which would probably be cleaner. That's out of scope for this PR though.